symmetric.go

```diff
@@ -0,0 +1,209 @@
+
+	// True if MAC supported
+	MAC bool
+
+	// ECB mechanism (CKM_..._ECB)
+	ECBMech uint
+}
+
+// CipherAES describes the AES cipher. Use this with the
+// GenerateSecretKey... functions.
+var CipherAES = SymmetricCipher{
+	KeyType:   pkcs11.CKK_AES,
+	GenMech:   pkcs11.CKM_AES_KEY_GEN,
+	BlockSize: 16,
+	Encrypt:   true,
+	MAC:       false,
+	ECBMech:   pkcs11.CKM_AES_ECB,
+}
+
+// CipherDES3 describes the three-key triple-DES cipher. Use this with the
+// GenerateSecretKey... functions.
+var CipherDES3 = SymmetricCipher{
+	KeyType:   pkcs11.CKK_DES3,
+	GenMech:   pkcs11.CKM_DES3_KEY_GEN,
+	BlockSize: 8,
+	Encrypt:   true,
+	MAC:       false,
+	ECBMech:   pkcs11.CKM_DES3_ECB,
+}
+
+// Ciphers is a map of PKCS#11 key types (CKK_...) to symmetric cipher information.
+var Ciphers = map[int]*SymmetricCipher{
+	pkcs11.CKK_AES:  &CipherAES,
+	pkcs11.CKK_DES3: &CipherDES3,
+}
+
+// PKCS11SecretKey contains a reference to a loaded PKCS#11 symmetric key object.
+//
+// A *PKCS11SecretKey implements the cipher.Block interface, allowing it be used
+// as the argument to cipher.NewCBCDecrypter, etc.
+type PKCS11SecretKey struct {
+	PKCS11Object
+
+	// Symmetric cipher information
+	Cipher *SymmetricCipher
+}
+
+// Key generation -------------------------------------------------------------
+
+// GenerateSecretKey creates an secret key of given length and type.
+//
+// The key will have a random label and ID.
+func GenerateSecretKey(bits int, cipher *SymmetricCipher) (*PKCS11SecretKey, error) {
+	return GenerateSecretKeyOnSlot(defaultSlot, nil, nil, bits, cipher)
+}
+
+// GenerateSecretKeyOnSlot creates as symmetric key on a specified slot
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func GenerateSecretKeyOnSlot(slot uint, id []byte, label []byte, bits int, cipher *SymmetricCipher) (*PKCS11SecretKey, error) {
+	var k *PKCS11SecretKey
+	var err error
+	if err = ensureSessions(libHandle, slot); err != nil {
+		return nil, err
+	}
+	err = withSession(slot, func(session *PKCS11Session) error {
+		k, err = GenerateSecretKeyOnSession(session, slot, id, label, bits, cipher)
+		return err
+	})
+	return k, err
+}
+
+// GenerateSecretKeyOnSession creates a symmetric key of given type and
+// length, on a specified session.
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func GenerateSecretKeyOnSession(session *PKCS11Session, slot uint, id []byte, label []byte, bits int, cipher *SymmetricCipher) (key *PKCS11SecretKey, err error) {
+	// TODO refactor with the other key generation implementations
+	if label == nil {
+		if label, err = generateKeyLabel(); err != nil {
+			return nil, err
+		}
+	}
+	if id == nil {
+		if id, err = generateKeyLabel(); err != nil {
+			return nil, err
+		}
+	}
+	secretKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_SECRET_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, cipher.KeyType),
```

Comment:
I've found SoftHSM is occasionally fussy about templates containing (redundant) class and key type definitions. Doesn't appear to affect AES keys, but RSA keys get rejected. Might be a good idea to remove the attributes in case other HSMs/implementations have an issue.

According to the spec it should be fine, but.. you know.